File: 2020/9.py
if __name__ == "__main__":
    data = list(map(int, open("9.txt").read().split('\n')))
    invalidNumber = None
    preambleLength = 25
    for i in range(preambleLength, len(data)):
        preamble = data[i-preambleLength:i]
        flag = True
        for j in range(len(preamble)):
            if data[i] - preamble[j] in set(preamble):
                flag = False
                break
        if flag:
            invalidNumber = data[i]
            print("Part1:", invalidNumber)
            break

    for i in range(len(data)-1):
        res = [data[i],data[i+1]]
        j = i+2
        while sum(res) < invalidNumber and j < len(data):
            res.append(data[j])
            j += 1
        
        if sum(res) == invalidNumber:
            print("Part2:", min(res)+max(res))
            break

File: 2015 (Go)/9.go
package main

import (
	"fmt"
	"io/ioutil"
	"math"
	"strconv"
	"strings"
)

type Graph map[string]map[string]int

func allVisited(visited map[string]bool) bool {
	for k := range visited {
		if visited[k] == false {
			return false
		}
	}
	return true
}

func tsp(graph Graph, visited map[string]bool, currNode string, globalValue *int, currValue int, calculatingMax bool) {
	if visited[currNode] {
		return
	}
	newVisited := make(map[string]bool)
	for k := range visited {
		newVisited[k] = visited[k]
	}
	newVisited[currNode] = true
	if allVisited(newVisited) {
		if calculatingMax {
			*globalValue = int(math.Max(float64(*globalValue), float64(currValue)))
		} else {
			*globalValue = int(math.Min(float64(*globalValue), float64(currValue)))
		}
		return
	}

	neighbours := graph[currNode]
	for n, v := range neighbours {
		tsp(graph, newVisited, n, globalValue, currValue+v, calculatingMax)
	}
}

func main() {
	// Travelling Salesman Problem (Directed Graph)

	d, _ := ioutil.ReadFile("9.txt")
	data := strings.Split(string(d), "\r\n")
	graph := make(Graph)
	visited := make(map[string]bool)
	for _, line := range data {
		lineData := strings.Split(line, " = ")
		nodes := strings.Split(lineData[0], " to ")
		if graph[nodes[0]] == nil {
			graph[nodes[0]] = make(map[string]int)
		}
		if graph[nodes[1]] == nil {
			graph[nodes[1]] = make(map[string]int)
		}
		graph[nodes[0]][nodes[1]], _ = strconv.Atoi(lineData[1])
		graph[nodes[1]][nodes[0]], _ = strconv.Atoi(lineData[1])
		visited[nodes[0]] = false
		visited[nodes[1]] = false
	}
	minValue := math.MaxInt64
	for city := range visited {
		for k := range visited {
			visited[k] = false
		}
		tsp(graph, visited, city, &minValue, 0, false)
	}
	fmt.Println("Part1:", minValue)

	maxValue := math.MinInt64
	for city := range visited {
		for k := range visited {
			visited[k] = false
		}
		tsp(graph, visited, city, &maxValue, 0, true)
	}
	fmt.Println("Part2:", maxValue)
}
